### PR TITLE
Remote control component

### PR DESCRIPTION
**Pull request by Eskjjlj**
## About The Pull Request
This is the draft PR for the remote control component and unmanned vehicles.

## Why It's Good For The Game
Eventually it will add more variety to the marine arsenal with remotely controlled vehicles.

### CODE REVIEW

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	if(istype(target, /obj/vehicle/unmanned))
+		var/obj/vehicle/unmanned/T = target
+		T.handle_remotecontrol_onuserclickon(A, params)
+
+	return COMSIG_MOB_CANCEL_CLICKON //So that the click returns there and doesn't go on.
+
+
+/datum/component/remote_control/proc/toggle_remote_control(datum/source, mob/user)
+	if(!is_controlling)
+		src.user = user
```
Comment (Rohesie) [inline code/modules/vehicles/unmanned/remote_control.dm]: Do this in `remote_control_on()`
It should probably never be turned on without a user.

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	RegisterSignal(parent, COMSIG_REMOTE_CONTROL_TOGGLE, .proc/toggle_remote_control)
+
+/datum/component/remote_control/Destroy()
+	return ..()
+
+/datum/component/remote_control/proc/on_user_clickon(datum/source, atom/A, params)
+	if(A==parent)
+		return ..()
+	if(istype(target, /obj/vehicle/unmanned))
```
Comment (Rohesie) [inline code/modules/vehicles/unmanned/remote_control.dm]: This check should happen in `Initialize()`, not after every click.
You can save the `on_user_clickon` proc path to a variable on initialize, and register it during `remote_control_on`

File: code/modules/vehicles/unmanned/unmanned_vehicle.dm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/obj/vehicle/unmanned/relaymove(mob/user, direction)
+	if(direction in GLOB.diagonals)
+		return
+	if(world.time <= last_move_time + move_delay)
+		return
+	step(src, direction)
+
+
+/obj/vehicle/unmanned/proc/handle_remotecontrol_onuserclickon(atom/A, params)
```
Comment (Rohesie) [inline code/modules/vehicles/unmanned/unmanned_vehicle.dm]: This can be a component proc.

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/datum/component/remote_control
+	var/mob/user
+	var/atom/controlled
+	var/is_controlling = FALSE
+	var/click_proc
+
+
+/datum/component/remote_control/Initialize(atom/controlled)
```
Comment (spookydonut) [inline code/modules/vehicles/unmanned/remote_control.dm]: you can't move /atom only /atom/movable

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
+/datum/component/remote_control
+	var/mob/user
+	var/atom/controlled
+	var/is_controlling = FALSE
+	var/click_proc
```
Comment (spookydonut) [inline code/modules/vehicles/unmanned/remote_control.dm]:
```suggestion
	var/datum/callback/click_proc
```

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		remote_control_on(user)
+	else
+		remote_control_off()
+
+/datum/component/remote_control/proc/remote_control_on(mob/user)
+	src.user = user
+	is_controlling = TRUE
+	user.remote_control = controlled	//Movement inputs are handled by the controlled thing in relaymove()
+	user.reset_perspective(controlled)
+	RegisterSignal(parent, COMSIG_ITEM_AFTERATTACK, click_proc)
```
Comment (spookydonut) [inline code/modules/vehicles/unmanned/remote_control.dm]: no checks for a null click_proc, will likely cause a runtime when it attempts call(null)

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/datum/component/remote_control
+	var/mob/user
+	var/atom/controlled
```
Comment (spookydonut) [inline code/modules/vehicles/unmanned/remote_control.dm]:
```suggestion
	var/atom/movable/controlled
```

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	var/is_controlling = FALSE
+	var/click_proc
+
+
+/datum/component/remote_control/Initialize(atom/controlled)
+	. = ..()
+	src.controlled = controlled
```
Comment (spookydonut) [inline code/modules/vehicles/unmanned/remote_control.dm]:
```suggestion
	if(!ismovableatom(controlled))
		return COMPONENT_INCOMPATIBLE
	src.controlled = controlled
```

File: code/modules/vehicles/unmanned/remote_control.dm [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+
+/datum/component/remote_control/Initialize(atom/target)
+	. = ..()
+	src.target = target
+
+	RegisterSignal(parent, COMSIG_REMOTE_CONTROL_TOGGLE, .proc/toggle_remote_control)
+
+/datum/component/remote_control/Destroy()
+	return ..()
```
[the remark this answers is not in the document]
Comment (spookydonut) [reply code/modules/vehicles/unmanned/remote_control.dm]: yeah you dont need this destroy

**Pull request closed without merging**